src/python/pants/engine/addressable.py

```diff
@@ -5,38 +5,46 @@
 from collections.abc import MutableMapping, MutableSequence
 from dataclasses import dataclass
 from functools import update_wrapper
-from typing import Any, Set, Tuple, Type, Union
+from typing import Any, Sequence, Set, Tuple, Type
 
 from pants.base.exceptions import ResolveError
-from pants.base.specs import AddressSpec, FilesystemResolvedSpec
+from pants.base.specs import OriginSpec
 from pants.build_graph.address import Address, BuildFileAddress
 from pants.engine.objects import Collection, Resolvable, Serializable
 from pants.util.objects import TypeConstraintError
 
 
+def _assert_single_address(addresses: Sequence[Address]) -> None:
+  """Assert that exactly one address must be contained in the collection."""
+  if len(addresses) == 0:
+    raise ResolveError("No targets were matched.")
+  if len(addresses) > 1:
+    output = '\n  * '.join(address.spec for address in addresses)
+    raise ResolveError(
+      "Expected a single target, but was given multiple targets.\n\n"
+      f"Did you mean one of:\n  * {output}"
+    )
```

Comment:
This is simply generalized out from `Addresses.expect_single()` so that we may reuse it for `AddressesWithOrigin.expect_single()`.